docs(ad-api): add operation mode api (#161)
autowarefoundation/autoware-documentation: docs/design/autoware-interfaces/ad-api/list/api/operation/mode/state.md

# /api/operation/mode/state

- Method: notification
- Type: [autoware_ad_api_msgs/msg/OperationModeState](../../../../types/autoware_ad_api_msgs/msg/operation_mode_state.md)

## Description

Get the operation mode state. For details, see the [operation mode](./index.md).

## Message

| Name                   | Type   | Description                                              |
| ---------------------- | ------ | -------------------------------------------------------- |
| operation_mode         | uint16 | The selected command for Autoware control.               |

> | is_in_autoware_control | bool   | True if direct control by the driver is enabled.         |

Reviewer: - We can consider renaming this value.
- The description is old.

@TakaHoribe @yukkysaito @mitsudome-r What do you think is the best name?

```suggestion
| is_engaged | bool   | True if vehicle control by Autoware is enabled.         |
```

Author: Or change API name ?
- `enable_autoware_control`
- `disable_autoware_control`

Reviewer: > Or change API name ?

It might be good because the word `engage` is a bit unclear for newbies. :+1:

Reviewer 2: > vehicle control by Autoware is enabled

I think this is a good high level description.
If we need a lower level description, I propose the following
```suggestion
| is_engaged | bool   | True if the control commands generated by Autoware are being sent to the vehicle.         |
```

Reviewer: @maxime-clem Thanks for the comment!
The description should be updated at some time so that everyone can easily understand things, but what is important at this stage is the naming because it's difficult to be changed later, especially if it's the name of APIs.

So do you have any opinion on the status/command name of turning autonomous control on/off?

Reviewer 3: Apologies for the late response.  

I was originally in agreement with @kenji-miyake and @maxime-clem but have been thinking about this some more. I feel that a developer might see `is_engaged` out of context and be unclear about its meaning. For example, whilst working with AutonomouStuff I often had to clarify that "engaged" in Autoware talk meant drive-by-wire or autonomous mode was enabled.

The [Merriam-Webster definition of `engage`](https://www.merriam-webster.com/dictionary/engage) includes the following:
> : to interlock with : [MESH](https://www.merriam-webster.com/dictionary/mesh)
>     also : to cause (mechanical parts) to mesh
>     // engage the clutch

So I can see where `engage` came from originally, as it seems like a common automotive term. However, since there is no mechanical interlocking/meshing when switching from manual to autonomous driving, perhaps we should use `enabled` instead of `engaged`? 🤔 

The description also of this parameter also says `enabled`, so I think it might be more clear to use `is_autoware_enabled` instead.

Author: Change API name https://github.com/autowarefoundation/autoware-documentation/pull/161/commits/4250bfcc92178c373595f9b3b97ff947a5e3f189.

Author: @LalithVipulananthan  Thank you for the comment.

I prefer `is_under_autoware_control` because the API covers all Autoware features. Without the word `control`, users may not understand that it is for the control of the vehicle.

@kenji-miyake @mitsudome-r @yukkysaito What do you think?

Reviewer: Either `is_autoware_control_enabled` or `is_under_autoware_control` is okay for me.
Considering the API name `enable_autoware_control`, `is_autoware_control_enabled` might be easier to understand.

Regarding adding the word `control`, it is good to be more descriptive. :+1:

Author: Change field name https://github.com/autowarefoundation/autoware-documentation/pull/161/commits/2c8a0d78be1cb6c4a6aefe9237191e1a88b2f130.
